### PR TITLE
security-check: test for `_FORTIFY_SOURCE` usage in release binaries

### PR DESCRIPTION
**Pull request by fanquake**
Test for the existence of fortified functions in the ELF release binaries.
Currently skips `bitcoin-util` and checks for RISC-V.

### CODE REVIEW

**Comment by fanquake** [comment]
> we don't end up with any foritfied funcs in bitcoin-util or bitcoin-cli.

`bitcoin-cli` is fixed once we fortify libevent.

**Comment by fanquake** [comment]
> bitcoin-cli is fixed once we fortify libevent.

Rebased on #27118

**Comment by fanquake** [comment]
Rebased post #27118.

**Comment by hebasto** [comment]
Concept ACK.

**Comment by hebasto** [comment]
> Can't be done yet because we don't end up with any fortified funcs in `bitcoin-util`.

Does it make sense to provide a list of expected symbols for every binary been tested? It would be an empty one for `bitcoin-util`.

**Comment by fanquake** [comment]
> Does it make sense to provide a list of expected symbols for every binary been tested?

I don't think so, and that would likely require constant maintenance (plus guix builds to be run on every code change).

File: contrib/devtools/security-check.py [hunk 1]
```diff
@@ -116,6 +117,21 @@ def check_ELF_control_flow(binary) -> bool:
         return True
     return False
 
+def check_ELF_FORTIFY(binary) -> bool:
+
+    chk_funcs = set()
+
+    for sym in binary.symbols:
+        match = re.search(r'__[a-z]*_chk', sym.name)
```
Comment (laanwj) [inline contrib/devtools/security-check.py:125]: Might want to check `.imported` to make sure it's an imported symbol, just to be sure.

**Comment by laanwj** [comment]
i'd be okay with skipping the check for `bitcoin-util`: it's the least relevant binary for fortification (no network access, not even file format access). Could reconsider it later if it actually gains some useful functionality 😄

File: contrib/devtools/security-check.py (hunk 1 above)
Comment (fanquake) [reply contrib/devtools/security-check.py:129]: Added.

**Comment by fanquake** [comment]
> i'd be okay with skipping the check for bitcoin-util: it's the least relevant binary for fortification (no network access, not even file format access). Could reconsider it later if it actually gains some useful functionality 😄

I think I agree, and I've added that skipping now.

**Comment by fanquake** [comment]
RISCV build failed here:
```bash
Traceback (most recent call last):
  File "/distsrc-base/distsrc-3733b4b088ff-riscv64-linux-gnu/./contrib/devtools/test-security-check.py", line 73, in test_ELF
    self.assertEqual(call_security_check(cxx, source, executable, pass_flags + ['-no-pie','-fno-PIE']), (1, executable + ': failed PIE'))
AssertionError: Tuples differ: (1, 'test1: failed PIE FORTIFY') != (1, 'test1: failed PIE')

First differing element 1:
'test1: failed PIE FORTIFY'
'test1: failed PIE'

- (1, 'test1: failed PIE FORTIFY')
?                       --------

+ (1, 'test1: failed PIE')
```

**Comment by fanquake** [comment]
> RISCV build failed here:

This might be a bug in LIEF. Opened https://github.com/lief-project/LIEF/issues/1082 upstream. Pushed up a verison with a workaround for RISC-V.
Guix Build (aarch64):
```bash
71aadd4cdca388e39742911d0a9a8f858cc796d1037ff4ff05f383bf74b6dfc6  guix-build-573db37893d8/output/aarch64-linux-gnu/SHA256SUMS.part
723ab391cadeda9afd12d832db26c65f2d69e9e41c92b20a2453821e5fe13dc2  guix-build-573db37893d8/output/aarch64-linux-gnu/bitcoin-573db37893d8-aarch64-linux-gnu-debug.tar.gz
d04eed5f8d5fadca37c57d18dcd908b71fa66425ee2388227b018f8df7d7cf5e  guix-build-573db37893d8/output/aarch64-linux-gnu/bitcoin-573db37893d8-aarch64-linux-gnu.tar.gz
e7757d05fce2b0a36e6f5785f3f7dbc52985fd4baeb111fc6745ad302ebf1afd  guix-build-573db37893d8/output/arm-linux-gnueabihf/SHA256SUMS.part
9f614712ae2dfd1dd5a6f15d4dc96349bbf81372245ef7b835cc0366a4f87e16  guix-build-573db37893d8/output/arm-linux-gnueabihf/bitcoin-573db37893d8-arm-linux-gnueabihf-debug.tar.gz
46d7992241831d4a2a233ec1e984f93f95623672956e80b9bcf4a163ae6a9c9a  guix-build-573db37893d8/output/arm-linux-gnueabihf/bitcoin-573db37893d8-arm-linux-gnueabihf.tar.gz
4da3fc982733ac3b5b7214408483ae7626e696ce533862d990e4124c59511fcd  guix-build-573db37893d8/output/arm64-apple-darwin/SHA256SUMS.part
e394ea1b46ee992b7e91d397f4b307f4d9e51bf1019b9c184eae8edbd07ddd9f  guix-build-573db37893d8/output/arm64-apple-darwin/bitcoin-573db37893d8-arm64-apple-darwin-unsigned.tar.gz
602e8a4e3f2b9b55a55f9266e496230d0c049f8fbea1cbda074d2fb2cc0a7549  guix-build-573db37893d8/output/arm64-apple-darwin/bitcoin-573db37893d8-arm64-apple-darwin-unsigned.zip
2172545ab072c79b248e887652a0ea8f375f6907f92d0a593c2cdb57ad03462c  guix-build-573db37893d8/output/arm64-apple-darwin/bitcoin-573db37893d8-arm64-apple-darwin.tar.gz
a9639798be0797c05e05fb2a337ef69dcb64572456a6c9d754f309c4f4870745  guix-build-573db37893d8/output/dist-archive/bitcoin-573db37893d8.tar.gz
60b2c1abfda3981d973ffdc189892110d54d2d161e9dfc690bbbc32dd7d12397  guix-build-573db37893d8/output/powerpc64-linux-gnu/SHA256SUMS.part
dea8e5ae3401e8901775aa5a1b8bcbbd64967fe3dc2dc62174f8ad3bfc7f5c39  guix-build-573db37893d8/output/powerpc64-linux-gnu/bitcoin-573db37893d8-powerpc64-linux-gnu-debug.tar.gz
8c4e3561759f92e9e03fdf4305da1a42b58f71711a999f33064c39bfeca725a7  guix-build-573db37893d8/output/powerpc64-linux-gnu/bitcoin-573db37893d8-powerpc64-linux-gnu.tar.gz
9c35b41b09a70d843c6753f1e4d5e8d40c80bd17f9348b8e852a2142a07973ee  guix-build-573db37893d8/output/riscv64-linux-gnu/SHA256SUMS.part
77f84e572d8f931572f5d60e2d61c083232b09cf2232a73d05931664ced815c0  guix-build-573db37893d8/output/riscv64-linux-gnu/bitcoin-573db37893d8-riscv64-linux-gnu-debug.tar.gz
ef965579cbc9647d0c92b8200520e746b602769a0728c61c36621cae6af7a687  guix-build-573db37893d8/output/riscv64-linux-gnu/bitcoin-573db37893d8-riscv64-linux-gnu.tar.gz
161a95df9654dfee9a94500c5cc03696502a0d7455538cc498c08cbdb99c56b1  guix-build-573db37893d8/output/x86_64-apple-darwin/SHA256SUMS.part
1626b5d6691a6710d171090c4d8036c44c8585a0b3c2899a6ea30c6333d1e923  guix-build-573db37893d8/output/x86_64-apple-darwin/bitcoin-573db37893d8-x86_64-apple-darwin-unsigned.tar.gz
23bdd9f1da7be43f86ab47b889efe73dc7e32dd1bfd8a44a610a86b2e872d320  guix-build-573db37893d8/output/x86_64-apple-darwin/bitcoin-573db37893d8-x86_64-apple-darwin-unsigned.zip
f18231be4b667917a8a75ce539dbec3fcf2f3279abe3cbf69104ba90da0154ca  guix-build-573db37893d8/output/x86_64-apple-darwin/bitcoin-573db37893d8-x86_64-apple-darwin.tar.gz
660054c7cc08d9d10fbd15ec4e818ee0193a15ad5a274c464154e8ea4f122348  guix-build-573db37893d8/output/x86_64-linux-gnu/SHA256SUMS.part
da66029304a66a0534930b7afaa561354d6b147aed3eafe2e5c2a44959b04cf5  guix-build-573db37893d8/output/x86_64-linux-gnu/bitcoin-573db37893d8-x86_64-linux-gnu-debug.tar.gz
0ab1514591fd1b74e2089e67ff0ef9f81fe72dcfe5932eaf3e9b4b9f96d3d96f  guix-build-573db37893d8/output/x86_64-linux-gnu/bitcoin-573db37893d8-x86_64-linux-gnu.tar.gz
30cad6948053c905234e51996f6512b59faaedbe7e7eb186559611a22c186df4  guix-build-573db37893d8/output/x86_64-w64-mingw32/SHA256SUMS.part
4753f415470f2c816b8ae1be04e09705e51ab57129dc22c680525cf5bb26f40e  guix-build-573db37893d8/output/x86_64-w64-mingw32/bitcoin-573db37893d8-win64-debug.zip
80bf9135f34faec17b962d94b61cf62d5a118908d5544194a416823c57afe015  guix-build-573db37893d8/output/x86_64-w64-mingw32/bitcoin-573db37893d8-win64-setup-unsigned.exe
516d5499655c57cc07cb7df67bbbd4d4d10260bdfa838229e33ba627cf2c3442  guix-build-573db37893d8/output/x86_64-w64-mingw32/bitcoin-573db37893d8-win64-unsigned.tar.gz
d5f4ced701f65bddd6ac46621d295d8c98a6ff6d73584452fa9d7ddfa14481cb  guix-build-573db37893d8/output/x86_64-w64-mingw32/bitcoin-573db37893d8-win64.zip
```

**Comment by fanquake** [comment]
> This might be a bug in LIEF. Opened https://github.com/lief-project/LIEF/issues/1082 upstream.

I haven't yet tested, but the bug should now be fixed, as of https://github.com/lief-project/LIEF/commit/ab85865f279cf02648018417ec8afa12bd0bef24.

**Comment by fanquake** [comment]
Guix Build (aarch64):
```bash
f21ba8dde52ecf7fd1e76b5f6efe7d620a7a0baf8099eb6d25d09161a73f2030  guix-build-6c9000cfbfab/output/aarch64-linux-gnu/SHA256SUMS.part
5e89a83e025bfbc37118a1ca0d386b0ffbde22bd3357b06d7e4c3587e10b0ed6  guix-build-6c9000cfbfab/output/aarch64-linux-gnu/bitcoin-6c9000cfbfab-aarch64-linux-gnu-debug.tar.gz
d0b7acb03204f077bc88d0d9008041c1aa854a6e0fd244210099ae2f2b8b4ca3  guix-build-6c9000cfbfab/output/aarch64-linux-gnu/bitcoin-6c9000cfbfab-aarch64-linux-gnu.tar.gz
e4619881f66bbefd26cc4e28723c3ebb9b6720f4a3f83ee4733e41246a72eff6  guix-build-6c9000cfbfab/output/arm-linux-gnueabihf/SHA256SUMS.part
0d3fcb044b41e7b9146329f9a9a2113fd55acab2f2129d58c2b56fca2cf08b0d  guix-build-6c9000cfbfab/output/arm-linux-gnueabihf/bitcoin-6c9000cfbfab-arm-linux-gnueabihf-debug.tar.gz
0dde42354c7322480a43281655c6a337429522b6cf845f8dcf154eb87cb4077d  guix-build-6c9000cfbfab/output/arm-linux-gnueabihf/bitcoin-6c9000cfbfab-arm-linux-gnueabihf.tar.gz
42b54602d398545561dc2fdc9d8ba255333a24a1c3d61e7455e79c1ac89d0388  guix-build-6c9000cfbfab/output/arm64-apple-darwin/SHA256SUMS.part
862b895a04b29266e8677d12ec3ed86c90d697e565bfe09b60e7fd987579ddb0  guix-build-6c9000cfbfab/output/arm64-apple-darwin/bitcoin-6c9000cfbfab-arm64-apple-darwin-unsigned.tar.gz
f30b60f67b6415c6598a64aeb64f5653349e9dfdeacb901c4173a73fb725904f  guix-build-6c9000cfbfab/output/arm64-apple-darwin/bitcoin-6c9000cfbfab-arm64-apple-darwin-unsigned.zip
8024ed97ffb3d639882b399dca0c8e305627663d25bc7afc0a3b408a2054590f  guix-build-6c9000cfbfab/output/arm64-apple-darwin/bitcoin-6c9000cfbfab-arm64-apple-darwin.tar.gz
f69554b8a178d80980a0235bb1380dfc7bfff51bffc8d5861540623c26f2088d  guix-build-6c9000cfbfab/output/dist-archive/bitcoin-6c9000cfbfab.tar.gz
e3fe7b51ae096171b0c1f5cadc3f81809b704dc443f4cdcbfdf4da314b308367  guix-build-6c9000cfbfab/output/powerpc64-linux-gnu/SHA256SUMS.part
4ed6abfc07cee15820ef0b82cf627d743eda7c1bf949f91a7259adbfcc908972  guix-build-6c9000cfbfab/output/powerpc64-linux-gnu/bitcoin-6c9000cfbfab-powerpc64-linux-gnu-debug.tar.gz
72c28fe0d63ca94df4629b05530576f432fa170d5dff9117c7e1fe177df1e33e  guix-build-6c9000cfbfab/output/powerpc64-linux-gnu/bitcoin-6c9000cfbfab-powerpc64-linux-gnu.tar.gz
36efec196510892ae01bfbae75d6fd624be97fba9cdfe98a84f4b743c58e7914  guix-build-6c9000cfbfab/output/riscv64-linux-gnu/SHA256SUMS.part
713afe90c3fbb54d42d16636b4cc66f94956498e043580510cb0cbcbf1cb0b72  guix-build-6c9000cfbfab/output/riscv64-linux-gnu/bitcoin-6c9000cfbfab-riscv64-linux-gnu-debug.tar.gz
9dc73a4f7fabf55853cb55bcdb7dc64b4670345d2c00d40aea827c62e8dcb8f6  guix-build-6c9000cfbfab/output/riscv64-linux-gnu/bitcoin-6c9000cfbfab-riscv64-linux-gnu.tar.gz
5d059bb7188b104b530c6a21108f94942d6f7acaabf469f1205e3a815721449e  guix-build-6c9000cfbfab/output/x86_64-apple-darwin/SHA256SUMS.part
e7a2b21f3f0737cc5a04a0e4cb7b64afc254347758be97f6312700e85932de2a  guix-build-6c9000cfbfab/output/x86_64-apple-darwin/bitcoin-6c9000cfbfab-x86_64-apple-darwin-unsigned.tar.gz
a45a52a60dd218cae67faaca037d4004da8b1ba299b04e80780d5a416ed6596a  guix-build-6c9000cfbfab/output/x86_64-apple-darwin/bitcoin-6c9000cfbfab-x86_64-apple-darwin-unsigned.zip
460ea30d33f604ed044b9ee66dfbecbc0fb7049a18a899d365420d8c3bc31d35  guix-build-6c9000cfbfab/output/x86_64-apple-darwin/bitcoin-6c9000cfbfab-x86_64-apple-darwin.tar.gz
6dca6e13803d9ea07c20e9e865b48aae4e5e5e7ac0e0dd6f5699990769e165f3  guix-build-6c9000cfbfab/output/x86_64-linux-gnu/SHA256SUMS.part
e4be9ef51c5fecfbf6bdf31508dc8f6aaaf31d46d49101da2b3a2e86c30513d8  guix-build-6c9000cfbfab/output/x86_64-linux-gnu/bitcoin-6c9000cfbfab-x86_64-linux-gnu-debug.tar.gz
686499d3990e762945279536aa9c84e6c79fc8e85d1e4ecedbb0ce9c70c3154c  guix-build-6c9000cfbfab/output/x86_64-linux-gnu/bitcoin-6c9000cfbfab-x86_64-linux-gnu.tar.gz
efe2376411ac6979dfe6e542b4bd1d5e950f069ba057755cf3a1cfac9aa3b619  guix-build-6c9000cfbfab/output/x86_64-w64-mingw32/SHA256SUMS.part
b3fcd800835eea1430e19f7cab82006c96f15b86985d9582d81e62de3fe6c195  guix-build-6c9000cfbfab/output/x86_64-w64-mingw32/bitcoin-6c9000cfbfab-win64-debug.zip
2b8e9e6cc2a0f2215d21599db5b03d2c9f7a158f3182908682b23df4b643976f  guix-build-6c9000cfbfab/output/x86_64-w64-mingw32/bitcoin-6c9000cfbfab-win64-setup-unsigned.exe
3e476530d1321acb2561027ce8ea05fc9a98430c6afc6a50ccd60ace72445f5b  guix-build-6c9000cfbfab/output/x86_64-w64-mingw32/bitcoin-6c9000cfbfab-win64-unsigned.tar.gz
bb55ef5ebc3e768899c71033dae45fbb45e26c415209047f460d5d6f3e98529c  guix-build-6c9000cfbfab/output/x86_64-w64-mingw32/bitcoin-6c9000cfbfab-win64.zip
```

**Comment by laanwj** [comment]
ACK 6c9000cfbfab1cd3b48efebd8e0e90ae597cf561

Building on x86_64 gets different guix output for MacOS. But this is unrelated to this PR.
```
-129b365bba906f50926218d5b8bb76921f5f637549fe44941e01eb8ba5d5d5f0  guix-build-6c9000cfbfab/output/arm64-apple-darwin/bitcoin-6c9000cfbfab-arm64-apple-darwin-unsigned.zip
+f30b60f67b6415c6598a64aeb64f5653349e9dfdeacb901c4173a73fb725904f  guix-build-6c9000cfbfab/output/arm64-apple-darwin/bitcoin-6c9000cfbfab-arm64-apple-darwin-unsigned.zip
-f836231d54d171cd98589519166200ed6a65741ff249e6e2144e94cf0e964ce5  guix-build-6c9000cfbfab/output/x86_64-apple-darwin/bitcoin-6c9000cfbfab-x86_64-apple-darwin-unsigned.zip
+a45a52a60dd218cae67faaca037d4004da8b1ba299b04e80780d5a416ed6596a  guix-build-6c9000cfbfab/output/x86_64-apple-darwin/bitcoin-6c9000cfbfab-x86_64-apple-darwin-unsigned.zip
```

File: contrib/devtools/security-check.py [hunk 2]
```diff
@@ -116,6 +117,25 @@ def check_ELF_CONTROL_FLOW(binary) -> bool:
...
+    chk_funcs = set()
+
+    for sym in binary.imported_symbols:
+        match = re.search(r'__[a-z]*_chk', sym.name)
+        if match:
+            chk_funcs.add(match.group(0))
+
+    # ignore stack-protector and bdb
+    chk_funcs.discard('__stack_chk')
+    chk_funcs.discard('__db_chk')
```
Comment (TheCharlatan) [inline contrib/devtools/security-check.py:135]: Bit unfortunate that this list has to be maintained, but I can't think of a better way either.

**Review by sedited — APPROVED** [review]
ACK be4f78275fa6608b11377dd5a29a809597d3fe8d

**Comment by hebasto** [comment]
Shouldn't we check for the absence of unfortified versions of functions, rather than checking for the presence of fortified ones?

**Comment by fanquake** [comment]
> Shouldn't we check for the absence of unfortified versions of functions, rather than checking for the presence of fortified ones?

Can you explain how that test would work? Not all functions are gauranteed to be fortified.

File: contrib/devtools/security-check.py (hunk 2 above)
Comment (fanquake) [reply contrib/devtools/security-check.py:135]: Yea, it's a bit annoying. At least post-BDB, it'll just be "ignore the stack protector".

**Comment by hebasto** [comment]
> > Shouldn't we check for the absence of unfortified versions of functions, rather than checking for the presence of fortified ones?
> 
> Can you explain how that test would work? Not all functions are guaranteed to be fortified.

1. Get the list of fortified functions from all binaries.
2. Create a list of the corresponding unfortified functions based on the list from step 1.
3. Check for the absence of symbols from the list in step 2.

**Comment by fanquake** [comment]
> Get the list of fortified functions from all binaries.

This assumes that fortification is already working correctly, otherwise you'll miss any (relevant) function that hasn't be fortified at least one time.

Again, it's not a bug to have unfortified functions. So I'm not sure what you are trying to acheive by turning that into a check failure. Can you explain further.

**Comment by hebasto** [comment]
> Again, it's not a bug to have unfortified functions.

Then how it can be classified?

> So I'm not sure what you are trying to acheive by turning that into a check failure.

Fortified functions can be statically linked into executables from a toolchain that was built with fortification enabled, while the rest of the code remains unfortified. In that case, the current PR branch will report a false positive result, won't it?

**Comment by fanquake** [comment]
> Then how it can be classified?

As expected behaviour, given that whether fortification occurs is dependant on various (compiler & libc) heuristics.

> Fortified functions can be statically linked into executables from a toolchain that was built with fortification enabled, while the rest of the code remains unfortified. In that case, the current PR branch will report a false positive result, won't it?

I don't really understand what you mean. Can you give a specific example of a false positive/issue, in the context of our Guix environment/build.

**Comment by hebasto** [comment]
> Can you give a specific example of a false positive/issue, in the context of our Guix environment/build.

Sure. This [branch](https://github.com/hebasto/bitcoin/commits/pr27038/0910-DEMO.2/) clearly demonstrates false positive results for `bitcoind` and `bitcoin-qt`:
```
b07d7f1b7e5f5eaf8649685e7f8e031e4ac078f87dbd27e0d732c2a578ef3c4c  guix-build-423fc912bca9/output/dist-archive/bitcoin-423fc912bca9.tar.gz
e7cce3c0bdf87e4583067fdf39aea50fb4c12fb0d52cbb78e3c7c0c967a9b215  guix-build-423fc912bca9/output/x86_64-linux-gnu/SHA256SUMS.part
b14a23e900a0ed33a4cc6ca274070e793bb0c649155f2ecac236f3c58438c06c  guix-build-423fc912bca9/output/x86_64-linux-gnu/bitcoin-423fc912bca9-x86_64-linux-gnu-debug.tar.gz
a4e71897a197c0c27fa303cd701278ae2c2354e13db78bb7f74fa9ee191ab01a  guix-build-423fc912bca9/output/x86_64-linux-gnu/bitcoin-423fc912bca9-x86_64-linux-gnu.tar.gz
```

**Comment by fanquake** [comment]
> Sure. This branch clearly demonstrates false positive results for bitcoind and bitcoin-qt:

Thanks. I built this branch and inspected `bitcoind`, and it contains calls to fortified functions. i.e:
```bash
objdump -D /root/bitcoin/guix-build-423fc912bca9/output/x86_64-linux-gnu/bitcoin-423fc912bca9/bin/bitcoind
<snip>
  6f9afe:	e8 cd dd 95 ff       	call   578d0 <__vsnprintf_chk@plt>
  6fd0a1:	e8 ba b1 95 ff       	call   58260 <__fprintf_chk@plt>
  6fe135:	e8 66 a0 95 ff       	call   581a0 <__fdelt_chk@plt>
  6fe161:	e8 3a a0 95 ff       	call   581a0 <__fdelt_chk@plt>
  6fe1ec:	e8 af 9f 95 ff       	call   581a0 <__fdelt_chk@plt>
  6fe228:	e8 73 9f 95 ff       	call   581a0 <__fdelt_chk@plt>
  6fe410:	e8 8b 9d 95 ff       	call   581a0 <__fdelt_chk@plt>
  6fe461:	e8 3a 9d 95 ff       	call   581a0 <__fdelt_chk@plt>
  6fe47f:	e8 1c 9d 95 ff       	call   581a0 <__fdelt_chk@plt>
  6ff600:	e8 1b 86 95 ff       	call   57c20 <__stack_chk_fail@plt>
  6ff85c:	e8 6f 8a 95 ff       	call   582d0 <__memcpy_chk@plt>
  6ff951:	e8 ca 82 95 ff       	call   57c20 <__stack_chk_fail@plt>
<much more output>
```

So how is this a false positive?

**Comment by hebasto** [comment]
> So how is this a false positive?

This [commit](https://github.com/hebasto/bitcoin/commit/a4b627a27867b274cc7541336e0995ffad58c632) deletes the source fortification logic from the build system altogether. Nevertheless, the check passes.

**Comment by fanquake** [comment]
> This [commit](https://github.com/hebasto/bitcoin/commit/a4b627a27867b274cc7541336e0995ffad58c632) deletes the source fortification logic from the build system altogether. Nevertheless, the check passes.

Yes, the check as implemented, which checks for (any) usage of foritfied function calls in the binary, passes, because the binary contains calls to fortified functions.

**Pull request merged**